### PR TITLE
fix(stepfunctions): task token integration cannot be used with API Gateway

### LINKED ISSUE #14181: (@aws-cdk/aws-stepfunction-tasks): Unnecessary validation of field "Headers" in CallApiGatewayRestApiEndpoint
The [validation](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-stepfunctions-tasks/lib/apigateway/base.ts#L31-L35) of Header property in constructor of CallApiGatewayRestApiEndpoint is unnecessary as StepFunction UI (i.e Amazon States Language) allows `apigateway:invoke.waitForTaskToken` without "Header" property. 

Also, as per HttpHeaderDeserializer in StepFunctionsHttpConnectors , Headers property is deserialized as `Map<String, List<String>>` . Constructing this structure in CDK with JsonPath is currently not possible. 

This validation is enforcing to add  `headers: TaskInput.fromObject({ TaskToken: JsonPath.taskToken)` in CallApiGatewayRestApiEndpoint props. When task with this configuration is deployed and run in AWS StepFunction it throws errror `parameter could not be used to start the Task: [The value of the field 'Headers' has an invalid format]"`

### Reproduction Steps

create Task, and deploy it. Try to run this stepFunction.

```
const task = new CallApiGatewayRestApiEndpoint(this, title, {
            api: restAPI,
            method: HttpMethod.PUT,
            stageName: apiStage,
            requestBody: TaskInput.fromObject({})
            authType: AuthType.IAM_ROLE,
            apiPath: apiPath,
            resultPath: resultPath,
            headers: TaskInput.fromObject({ TaskToken: JsonPath.taskToken),
            integrationPattern: IntegrationPattern.WAIT_FOR_TASK_TOKEN
          })

```
### What did you expect to happen?

Task should make a call to the ApiGateway

### What actually happened?

Exception during task execution 
`parameter could not be used to start the Task: [The value of the field 'Headers' has an invalid format]"`

### Environment

  - **CDK CLI Version  : 1.96.0**
  - **Framework Version: 1.96.0**
  - **Node.js Version: v14.6.0**
  - **OS               : MacOS**
  - **Language (Version): TypeScript** 

### Other

--- 

This is :bug: Bug Report

### LINKED ISSUE #14184: "Calling API gateway using Stepfunctions" feature is not properly supported by CDK, when using CallApiGatewayRestApiEndpoint
We are trying to use the [CallApiGatewayRestApiEndpoint](https://docs.aws.amazon.com/cdk/api/latest/python/aws_cdk.aws_stepfunctions_tasks/CallApiGatewayRestApiEndpoint.html) construct with WAIT_FOR_TASK_TOKEN integration pattern. Per the [AWS docs on the contract](https://docs.aws.amazon.com/step-functions/latest/dg/connect-api-gateway.html#connect-api-gateway-patterns), we need to specify the headers like this:

`"Headers": { 
            "TaskToken.$": "States.Array($$.Task.Token)"
        }`
        
However, we're not sure how to do that with the CallApiGatewayRestApiEndpoint construct. We tried something like this, which would generate the expected headers:

```
headers = aws_stepfunctions.TaskInput.from_object({
            "TaskToken.$":"States.Array($$.Task.Token)"
        })
step1 = aws_stepfunctions_tasks.CallApiGatewayRestApiEndpoint(
            self.wf.stack, "APIGW",
            api=rest_api,
            method=HttpMethod.POST,
            stage_name="prod",
            headers=headers,
            integration_pattern=aws_stepfunctions.IntegrationPattern.REQUEST_RESPONSE,
            auth_type=aws_stepfunctions_tasks.AuthType.IAM_ROLE
        )
```

but when we set IntegrationPattern to WAIT_FOR_TASK_TOKEN, CDK complains that task token is missing:

`Task Token is required in headers for WAIT_FOR_TASK_TOKEN pattern. Use JsonPath.taskToken to set the token.
`

But if we make the value of the header field a string instead of list/array, then CDK wouldn't raise an error. however when we execute this state machine, step functions would complain that the format of header field is incorrect. 

```
headers = aws_stepfunctions.TaskInput.from_object({
            "TaskToken.$":"$$.Task.Token"
        })
        step1 = aws_stepfunctions_tasks.CallApiGatewayRestApiEndpoint(
            self.wf.stack, "APIGW",
            api=rest_api,
            method=HttpMethod.POST,
            stage_name="prod",
            headers=headers,
            integration_pattern=aws_stepfunctions.IntegrationPattern.WAIT_FOR_TASK_TOKEN,
            auth_type=aws_stepfunctions_tasks.AuthType.IAM_ROLE
        )
```

### Reproduction Steps

Described above

### What did you expect to happen?

CDK shouldn't complain when we pass Task Token using States.Array within headers and when we use WAIT_FOR_TASK_TOKEN pattern.

### What actually happened?

Seems like for calling API gateway using stepfunctions, we need to pass task token within array inside the header. But CDK doesn't detect task token if we do this way. Also, note that CDK currently doesn't support intrinsic functions (Such as States.Array).

### Environment

  - **CDK CLI Version  :** 1.98.0
  - **Framework Version:** 1.98.0
  - **Node.js Version:** v14.6.0
  - **OS               :** Linux
  - **Language (Version):** Python (3.7.3)

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by rix0rrr**
To pass the Task Token in headers to an API Gateway, the token must
be wrapped in an array (because that's the value type of headers).

Because JSONPath evaluation needs to happen to resolve the token,
we need to use the `States.Array()` function in a `JsonPathToken`
to properly resolve this. However, doing that makes the existing
validation code fail the validation checking that you are passing
the task token somewhere.

Add convenience methods for the intrinsics, and update the checker
to also find paths referenced inside intrinsic functions.

Fixes #14184, fixes #14181.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-stepfunctions-tasks/README.md [hunk 1]
```diff
@@ -305,6 +305,25 @@ const invokeTask = new tasks.CallApiGatewayRestApiEndpoint(this, 'Call REST API'
 });
 ```
 
+Be aware that the header values must be arrays. When passing the Task Token
```
Comment (kaizen3031593) [inline packages/@aws-cdk/aws-stepfunctions-tasks/README.md:308]: I feel like the native intrinsic function support is a feature in itself and deserves to be mentioned in the readme too.

**Review by kaizen3031593 — CHANGES_REQUESTED** [review]
Looks good; some nits, mostly questions. And I think the intrinsics deserves mentioning in the readme.

File: packages/@aws-cdk/aws-stepfunctions/lib/private/intrinstics.ts [hunk 2]
```diff
@@ -0,0 +1,246 @@
+export type IntrinsicExpression = StringLiteralExpression | PathExpression | FnCallExpression;
+export type TopLevelIntrinsic = PathExpression | FnCallExpression;
```
[the remark this answers is not in the document]
Comment (rix0rrr) [reply packages/@aws-cdk/aws-stepfunctions/lib/private/intrinstics.ts:2]: Yes.

File: packages/@aws-cdk/aws-stepfunctions/lib/private/intrinstics.ts [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+    this.ws();
+
+    const ret =
+      this.char() === '$'
+        ? this.parsePath()
+        : isAlphaNum(this.char())
+          ? this.parseFnCall()
+          : this.raiseError("expected '$' or a function call");
+
+    this.ws();
```
[the remark this answers is not in the document]
Comment (rix0rrr) [reply packages/@aws-cdk/aws-stepfunctions/lib/private/intrinstics.ts:36]: `ws` gets called at any place where whitespace may occur, and moves the cursor to the next non-whitespace character.

File: packages/@aws-cdk/aws-stepfunctions/lib/private/intrinstics.ts [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+
+    if (!this.eof) {
```
[the remark this answers is not in the document]
Comment (rix0rrr) [reply packages/@aws-cdk/aws-stepfunctions/lib/private/intrinstics.ts:38]: What this checks is that after parsing exactly one expression (and potentially whitespace), we're at the end of the input.

If not, we got trailing crap that we cannot parse.

File: packages/@aws-cdk/aws-stepfunctions/lib/private/intrinstics.ts [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+      this.raiseError('unexpected trailing characters');
+    }
+
+    return ret;
+  }
+
+  public parseIntrinsic(): IntrinsicExpression {
```
[the remark this answers is not in the document]
Comment (rix0rrr) [reply packages/@aws-cdk/aws-stepfunctions/lib/private/intrinstics.ts:45]: It gets used in `parseFnCall`.

File: packages/@aws-cdk/aws-stepfunctions/lib/private/intrinstics.ts [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+      return this.parsePath();
+    }
+
+    if (isAlphaNum(this.char())) {
+      return this.parseFnCall();
+    }
+
+    if (this.char() === "'") {
+      return this.parseStringLiteral();
+    }
```
[the remark this answers is not in the document]
Comment (rix0rrr) [reply packages/@aws-cdk/aws-stepfunctions/lib/private/intrinstics.ts:58]: Limitation of underlying system.

**Review by kaizen3031593 — APPROVED** [review]

**Pull request merged**